File: packages/lynx-audio/ios/AudioEventBus.swift
import Foundation

/// Process-wide pub/sub for audio events (`onEnd`, `onMeter`) that the JS
/// shim subscribes to via `GlobalEventEmitter`. Mirrors the pattern from
/// `@sigx/lynx-notifications`/`@sigx/lynx-websocket` — a global bus
/// decoupled from any specific LynxView, plus a per-LynxView publisher
/// (`AudioPublisher`) that pumps each emission into JS through
/// `LynxView.sendGlobalEvent(name, withParams:)`.
final class AudioEventBus {

    static let shared = AudioEventBus()

    typealias Payload = [String: Any]
    typealias Listener = (String, Payload) -> Void

    private let queue = DispatchQueue(label: "com.sigx.audio.bus")
    private var listeners: [(token: UUID, fn: Listener)] = []

    @discardableResult
    func addListener(_ fn: @escaping Listener) -> UUID {
        let token = UUID()
        queue.sync { listeners.append((token, fn)) }
        return token
    }

    func removeListener(_ token: UUID) {
        queue.sync { listeners.removeAll { $0.token == token } }
    }

    func publishPlayerEnd(id: Int64) {
        emit(channel: "__sigxAudioEnd:\(id)", payload: ["id": id])
    }

    func publishMeter(id: Int64, peak: Double, avg: Double) {
        emit(channel: "__sigxAudioMeter:\(id)", payload: [
            "id": id,
            "peak": peak,
            "avg": avg,
        ])
    }

    private func emit(channel: String, payload: Payload) {
        let snapshot = queue.sync { listeners }
        for (_, fn) in snapshot { fn(channel, payload) }
    }
}

File: packages/lynx-audio/android/com/sigx/audio/AudioEventBus.kt
package com.sigx.audio

import com.lynx.react.bridge.JavaOnlyMap
import java.util.UUID
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Process-wide pub/sub for audio events. Players + recorders publish here;
 * per-LynxView [AudioPublisher] instances read and forward via
 * `LynxView.sendGlobalEvent(channel, [...])`.
 *
 * Each emission carries both a channel name (e.g. `__sigxAudioEnd:42`) and
 * a JSON-encodable payload. The JS side subscribes on per-id channels so
 * the JS shim can demux without parsing the payload first.
 */
internal object AudioEventBus {

    private val listeners = CopyOnWriteArrayList<Pair<UUID, (String, JavaOnlyMap) -> Unit>>()

    fun addListener(fn: (String, JavaOnlyMap) -> Unit): UUID {
        val token = UUID.randomUUID()
        listeners.add(token to fn)
        return token
    }

    fun removeListener(token: UUID) {
        listeners.removeAll { it.first == token }
    }

    fun publishPlayerEnd(id: Long) {
        val payload = JavaOnlyMap()
        payload.putDouble("id", id.toDouble())
        emit("__sigxAudioEnd:$id", payload)
    }

    fun publishMeter(id: Long, peak: Double, avg: Double) {
        val payload = JavaOnlyMap()
        payload.putDouble("id", id.toDouble())
        payload.putDouble("peak", peak)
        payload.putDouble("avg", avg)
        emit("__sigxAudioMeter:$id", payload)
    }

    private fun emit(channel: String, payload: JavaOnlyMap) {
        for ((_, fn) in listeners) fn(channel, payload)
    }
}
